robosuite/controllers/base_controller.py

```diff
@@ -0,0 +1,155 @@
+                 sim,
+                 id_name,
+                 joint_indexes,
+    ):
+
+        # Attributes for scaling / clipping inputs to outputs
+        self.action_scale = None
+        self.action_input_transform = None
+        self.action_output_transform = None
+
+        # Private property attributes
+        self._control_dim = None
+        self._output_min = None
+        self._output_max = None
+        self._input_min = None
+        self._input_max = None
+
+        # mujoco simulator state
+        self.sim = sim
+        self.model_timestep = self.sim.model.opt.timestep
+        self.id_name = id_name
+        self.joint_index = joint_indexes
+
+        # robot states
+        self.ee_pos = None
+        self.ee_ori_mat = None
+        self.ee_pos_vel = None
+        self.ee_ori_vel = None
+        self.joint_pos = None
+        self.joint_vel = None
+
+        # dynamics and kinematics
+        self.J_pos = None
+        self.J_ori = None
+        self.J_full = None
+        self.mass_matrix = None
+
+        # Joint dimension
+        self.joint_dim = len(joint_indexes)
+
+        # Torques being outputted by the controller
+        self.torques = None
+
+        # Move forward one timestep to propagate updates before taking first update
+        self.sim.forward()
+
+        # Initialize controller by updating internal state and setting the initial joint, pos, and ori
+        self.update()
+        self.initial_joint = self.joint_pos
+        self.initial_ee_pos = self.ee_pos
+        self.initial_ee_ori_mat = self.ee_ori_mat
+        self.torque_compensation = np.zeros(self.joint_dim)
+
+    @abc.abstractmethod
+    def run_controller(self, action):
+        """
+        Go from actions to torques
+        """
+        pass
+
+    def scale_action(self, action):
+        """
+        Scale the action based on max and min of action
+        """
+
+        if self.action_scale is None:
+            self.action_scale = abs(self.output_max - self.output_min) / abs(self.input_max - self.input_min)
+            self.action_output_transform = (self.output_max + self.output_min) / 2.0
+            self.action_input_transform = (self.input_max + self.input_min) / 2.0
+        action = np.clip(action, self.input_min, self.input_max)
+        transformed_action = (action - self.action_input_transform) * self.action_scale + self.action_output_transform
+
+        return transformed_action
+
+    def update(self):
+
+        self.ee_pos = np.array(self.sim.data.body_xpos[self.sim.model.body_name2id(self.id_name)])
+        self.ee_ori_mat = np.array(self.sim.data.body_xmat[self.sim.model.body_name2id(self.id_name)].reshape([3, 3]))
+        self.ee_pos_vel = np.array(self.sim.data.body_xvelp[self.sim.model.body_name2id(self.id_name)])
+        self.ee_ori_vel = np.array(self.sim.data.body_xvelr[self.sim.model.body_name2id(self.id_name)])
+
+        self.joint_pos = np.array(self.sim.data.qpos[self.joint_index])
```

Comment:
@cremebrule This might be using the wrong set of indices - you are using `env._ref_joint_vel_indexes` but we might need `env._ref_joint_pos_indexes` to index into the qpos array. Can you please check?